### DISCUSSION TITLE
Question about X Orthographic Zoom Aspect Ratio Preservation

### DISCUSSION THREAD

**Discussion by cornhundred** [visgl/deck.gl#7489] [Q&A] [answered]
Hi, I have a question about the behavior of orthographic zoom in one axis only. I'm using the code below to allow zooming in the X axis only with `ini_zoom` set to `[-2, -2]`. The data being visualized is a grid of red squares and the X zooming causes the overall grid of squares to contract and expand in the X direction (like an accordion) as expected. However, the individual red squares themselves are not expanding or contracting in the X direction and appear to be restricted their original aspect ratio. Would anyone know how to go about modifying the behavior to allow the individual red squares to 'stretch'?

### Stretching Behavior

![accordian_deckgl](https://user-images.githubusercontent.com/8352840/206305873-b0d5ce10-cb95-44a8-bb7e-4f846a56dae0.gif)
 
### Initialize deck
```
deckgl = {

  const instance = new deck.DeckGL({
    container,
    views: new deck.OrthographicView(),
    controller: {scrollZoom: true, inertia: true, zoomAxis: 'X'},
    initialViewState: {target: ini_target, zoom: ini_zoom},
    
  })
  
  return instance;
}
```

### Custom Primitive Layer Definition
```
layer2 = {
  const {Layer} = deck;
  
  class ScatterplotLayer extends Layer {
    initializeState() {
      const {gl} = this.context;

      // Register attributes
      this.getAttributeManager().addInstanced({
        instancePositions: {
          size: 3,
          type: gl.DOUBLE,
          accessor: 'getPosition'
        },
        instanceRadius: {
          size: 1,
          accessor: 'getRadius',
          defaultValue: 1
        },
        instanceColors: {
          size: 4,
          normalized: true,
          type: gl.UNSIGNED_BYTE,
          accessor: 'getColor',
          defaultValue: [0, 0, 0, 255]
        }
      });
      
      // Save the model in layer state
      this.setState({
        model: getModel(gl)
      });
    }
    
    updateState() {
      // Retrieve the model from layer state
      this.state.model.setUniforms({
        // smoothRadius: this.props.smoothRadius
      });
    }
  }
  ScatterplotLayer.layerName = 'ScatterplotLayer';
  ScatterplotLayer.defaultProps = defaultProps;

  const layer = new ScatterplotLayer({
    id: `scatterplot-${Date.now()}`,
    data,
    getPosition: d => d.position,
    getRadius: d => d.size,
    getColor: d => d.color
  });
  
  deckgl.setProps({layers: [layer]});
  
  return layer;
}
```

### Additional Background

Link to the [ObservableHQ notebook](https://observablehq.com/d/bece83082b143f05)

I modified this example from this notebook https://observablehq.com/@pessimistress/deck-gl-tutorial-custom-primitive-layer so there might be something about how this notebook is defining the layer that is 'locking' in this aspect ratio.

**Comment by Pessimistress** [comment, marked as answer]
Your layer is calculating geometry size in the common space. To achieve the effect you want, the corners of each square should use an absolute position in the world space. See https://deck.gl/docs/developer-guide/coordinate-systems#concepts

```patch
-  vec3 offsetCommon = positions * project_size(instanceRadius);
-  vec3 positionCommon = project_position(instancePositions, instancePositions64Low);
+  vec3 positionCommon = project_position(instancePositions + positions * instanceRadius, instancePositions64Low);
-  gl_Position = project_common_position_to_clipspace(vec4(positionCommon + offsetCommon, 1.0));
+  gl_Position = project_common_position_to_clipspace(vec4(positionCommon, 1.0));
```

**Reply by cornhundred** [reply]
Thank you @Pessimistress for the explanation and link to the documentation. It is now working as expected in [this notebook](https://observablehq.com/d/5cfc03dc054194d3)

![accordian_deckgl_absolute](https://user-images.githubusercontent.com/8352840/206352736-0c123589-f69c-446f-a6dd-9c0c8918c3d5.gif)
